applications/nrf_desktop/src/modules/dvfs.c

```diff
@@ -137,39 +166,111 @@ static void handle_dvfs_error(int32_t err)
 		timeout = DVFS_RETRY_INITIALIZATION_TIMEOUT;
 	} else {
 		LOG_ERR("DVFS freq change returned with error: %d", err);
-		module_set_state(MODULE_STATE_ERROR);
-		module_state = STATE_ERROR;
-		cancel_dvfs_retry_work();
+		dvfs_fatal_error();
 		return;
 	}
 	(void) k_work_reschedule(&dvfs_retry.retry_work, timeout);
 }
 
-static void set_dvfs_freq(enum dvfs_frequency_setting target_freq)
+static uint32_t check_required_frequency(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(dvfs_freq_array) - 1; i++) {
+		if (dfvs_requests_state_bitmask & dvfs_freq_array[i].bitmask) {
+			return dvfs_freq_array[i].freq;
+		}
+	}
+
+	/* If no state is active, return the lowest frequency. */
+	return dvfs_freq_array[ARRAY_SIZE(dvfs_freq_array) - 1].freq;
+}
+
+static void dvfs_notify_cb(struct onoff_manager *srv,
+			   struct onoff_client *cli,
+			   uint32_t state,
+			   int res)
 {
-	int32_t ret = dvfs_service_handler_change_freq_setting(target_freq);
+	(void) k_work_submit(&dvfs_notify_work);
+}
+
+static void set_dvfs_freq(uint32_t target_freq)
+{
+	if (target_freq == current_freq) {
+		return;
+	}
+	int ret;
+
+	if (spec.frequency != 0) {
+		ret = nrf_clock_control_release(dvfs_clock_dev, &spec);
+		if (ret < 0) {
+			LOG_ERR("Failed to release requested clock specs, error: %d", ret);
+			dvfs_fatal_error();
+			return;
+		}
+	}
+
+	sys_notify_init_callback(&cli.notify, dvfs_notify_cb);
 
+	spec.frequency = target_freq;
+	ret = nrf_clock_control_request(dvfs_clock_dev, &spec, &cli);
 	if (ret) {
 		handle_dvfs_error(ret);
 	} else {
-		current_freq = target_freq;
-		LOG_INF("Have requested %s frequency",
-			get_dvfs_frequency_setting_name(target_freq));
-		cancel_dvfs_retry_work();
+		LOG_INF("Have requested %" PRIu32 " frequency", target_freq);
+		requested_freq = target_freq;
+		request_in_progress = true;
 	}
 }
 
-static enum dvfs_frequency_setting check_required_frequency(void)
+static void set_dvfs_based_on_frequency(void)
```

Comment:
Can we call it e.g. `dvfs_frequency_update`/`dvfs_frequency_align`? You actually set the frequency here now